[AMDGPU] SIPeepholeSDWA: Add REG_SEQUENCE support

frederik-h opened the pull request:
The REG_SEQUENCE instruction represents a copy of several registers to the subregisters of a target register. The "si-peephole-sdwa" pass currently cannot handle such instructions as operands of machine instructions that are considered for the conversion to SDWA.

This commit extends the SDWASrcOperand implementation to allow it to treat the use of a subregister of a REG_SEQUENCE in an SDWASrcOperand as if the source register of the copy had been used directly.

Fixes #130102

frederik-h committed:
[AMDGPU] SIPeepholeSDWA: Add REG_SEQUENCE support

The REG_SEQUENCE instruction represents a copy of several
registers to the subregisters of a target register.
The si-peephole-sdwa pass currently cannot handle
such instructions as operands of machine instructions
that are considered for the conversion to SDWA.

This commit extends the SDWASrcOperand implementation
to allow it to treat the use of a subregister of a
REG_SEQUENCE in an SDWASrcOperand as if the source
register of the copy had been used directly.

arsenm commented on llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp:
```diff
@@ -382,6 +390,97 @@ uint64_t SDWASrcOperand::getSrcMods(const SIInstrInfo *TII,
+/// which is copied from \p Op, i.e. the operand following
+/// \p Op in the operands of \p RegSeq, or nullopt if the
+/// the \p Op is not an operand of \p RegSeq.
+static std::optional<unsigned> regSequenceFindSubreg(const MachineInstr &RegSeq,
```
This patch is inventing ever so slightly different versions of helpers which exist in some form or another

arsenm commented on llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp:
```diff
@@ -382,6 +390,97 @@ uint64_t SDWASrcOperand::getSrcMods(const SIInstrInfo *TII,
+  unsigned SubReg = *regSequenceFindSubreg(RegSeq, Reg);
+
+  MachineInstr *SingleUse = nullptr;
+  for (MachineInstr &UseMI : MRI->use_nodbg_instructions(SeqReg))
+    for (auto &Op : UseMI.operands())
+      if (Op.isReg() && Op.getReg() == SeqReg && Op.getSubReg() == SubReg) {
+        if (SingleUse)
+          return nullptr;
+        SingleUse = &UseMI;
+      }
```
MachineRegisterInfo should have getOneNonDBGUse and getOneNonDBGUser, not sure how we don't have them. I think I've seen that floating around in an open PR recently

frederik-h replied:
Found it in this open PR: https://github.com/llvm/llvm-project/pull/125885. I included it here: https://github.com/llvm/llvm-project/pull/133087/commits/d6fbc1abe6a8fb8760aed399becb2064a408ca69 and used it to replaced the explicit search for the single use here.

arsenm commented on llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp:
```diff
@@ -278,11 +281,16 @@ static void copyRegOperand(MachineOperand &To, const MachineOperand &From) {
 }
 
+static bool isSameReg(const MachineOperand &Op, Register Reg) {
+  return Op.isReg() && Op.getReg() == Reg;
```
isReg can never be false for the copy-like contexts you're looking at

frederik-h replied:
I have removed the `isSameReg` overloads that I had previously introduced and replaced the use of this one by `Op.getReg() == Reg`.

arsenm commented on llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp:
```diff
@@ -278,11 +281,16 @@ static void copyRegOperand(MachineOperand &To, const MachineOperand &From) {
+  return Op.isReg() && Op.getReg() == Reg;
+}
+
+static bool isSameReg(const MachineOperand &Op, Register Reg, unsigned SubReg) {
```
Probably should just use TargetInstrInfo::RegSubRegPair

frederik-h replied:
This overload of `isSameReg` was just introduced for consistency and is not really necessary.

arsenm commented on llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp:
```diff
@@ -382,6 +390,97 @@ uint64_t SDWASrcOperand::getSrcMods(const SIInstrInfo *TII,
+
+  auto *End = RegSeq.operands_end();
+  // Operand pair at indices (i+1, i+2) is (register, subregister)
+  for (auto *It = RegSeq.operands_begin() + 1; It != End; It += 2) {
+    if (isSameReg(*It, Reg))
+      return (It + 1)->getImm();
```
Looking for an exact match of the register in the input is weird. Normal uses of reg_sequence look for compatibility of the subregister with the use index. 

Why does the specific register matter? Usually you're just looking up the chain of copies for the original 32-bit source register. Maybe it's worth extracting ValueTracker out of PeepholeOpt as a utility?

frederik-h replied:
> Looking for an exact match of the register in the input is weird. Normal uses of reg_sequence look for compatibility of the subregister with the use index.

I guess this is not a normal use of the instruction then. Consider the example that I provided in a comment in this PR:
```llvm
%6:vgpr_32 = V_AND_B32_e64 %2, %5
%7:vreg_64 = REG_SEQUENCE %6, %subreg.sub0, %3, %subreg.sub1
%8:vgpr_32, %9:sreg_64_xexec = V_ADD_CO_U32_e64 %4.sub0, %7.sub0, 0
```

In `SDWASrcOperand::potentialToConvert`, we are given the register `%6` and we look at its users. We want to return `%8` as the "PotentialMI"  into which the instruction defining `%6` can be folded. But since `%8` does not use `%6` directly, we only find `%7` as a user of `%6`. Then we use this function to find the subregister `sub0` of `%7` to which `%6`  gets copied. This is then used to find the "real PotentialMI" `%8` by looking for a use of `%7` which accesses `%7:sub0`.

> Why does the specific register matter? Usually you're just looking up the chain of copies for the original 32-bit source register. Maybe it's worth extracting ValueTracker out of PeepholeOpt as a utility?

I don't know. Could it be used for this "reverse" lookup, i.e. not  from `%7.sub0` to `%6` but the other way round?

arsenm commented on llvm/test/CodeGen/AMDGPU/sdwa-peephole-reg-sequence.mir:
```diff
@@ -0,0 +1,55 @@
+    ; CHECK-NEXT: [[COPY1:%[0-9]+]]:vreg_64 = COPY [[DEF]]
+    ; CHECK-NEXT: GLOBAL_STORE_DWORDX2 killed [[COPY1]], killed [[REG_SEQUENCE2]], 0, 0, implicit $exec :: (store (s64), addrspace 1)
+    ; CHECK-NEXT: S_ENDPGM 0
+    ; CHECK-NEXT: {{  $}}
+    ; CHECK-NEXT: .1.entry:
```
Where is this coming from?

frederik-h replied:
This was introduced by the script that I used for generating the checks somehow. But I have fixed the RUN line and regenerated the test (which I have now split into two test files) and now the weird check lines are gone.

arsenm commented on llvm/test/CodeGen/AMDGPU/sdwa-peephole-reg-sequence.mir:
```diff
@@ -0,0 +1,55 @@
+# NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 5
+# RUN: llc -mtriple=amdgcn-amd-amdhsa -mcpu=gfx900  -stop-after=si-peephole-sdwa -o - %s | FileCheck %s
```
I don't understand what this run line is doing, and it should probably be an llc error. Where is this starting? I'm suspecting it's forming the default empty function, codegenning that, and then it somehow merges with whatever MIR happened to be there from the initial parse?

frederik-h replied:
Thanks, fixed it. I am also not sure what it really does.

arsenm commented on llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp:
```diff
@@ -391,21 +490,23 @@ MachineInstr *SDWASrcOperand::potentialToConvert(const SIInstrInfo *TII,
+    // Check that all instructions that use Reg can be converted
+    for (MachineInstr &UseMI :
+         getMRI()->use_nodbg_instructions(Reg->getReg())) {
+      MachineInstr *SrcMI = findUseSrc(UseMI, *Reg, getMRI());
```
This is looking at the uses of Reg->getReg() inside a loop over the uses of Reg->getReg()?

frederik-h replied:
No, see also my other [comment](https://github.com/llvm/llvm-project/pull/133087/files/243157eb5f2437c6c8e6f9cecbbdc173a2e8c363#r2023045944); `regSequenceFindSingleSubregUse` does not iterate the uses of `Reg->getReg()` again (which is `%6` in the example from the other comment). It iterates the uses of the register defined by `UseMI` (`%7` in the example) if it is a `REQ_SEQUENCE`.

arsenm commented on llvm/test/CodeGen/AMDGPU/sdwa-peephole-reg-sequence.mir:
```diff
@@ -0,0 +1,55 @@
+    %13:sreg_64 = IMPLICIT_DEF
+    %14:vreg_64 = COPY %13
+    GLOBAL_STORE_DWORDX2 killed %14, killed %12, 0, 0, implicit $exec :: (store (s64), addrspace 1)
+    S_ENDPGM 0
```
Add some tests with multiple uses, and where you need to compose subregisters?

frederik-h replied:
> Add some tests [...] where you need to compose subregisters?
Right now the code form this PR does not support "chains of REG_SEQUENCEs", but it should be possible to support this if you think this is necessary.

frederik-h committed:
Add MachineRegisterInfo::getOneNonDBGUse{,r}

frederik-h committed:
Review changes

- Revert changes to isSameReg function (not necessary)
- Use new  MRI->getOneNonDBGUser instead of checking for single use
  manually
- Remove findUseSrc
- Split test/CodeGen/AMDGPU/sdwa-peephole-reg-sequence.mir into two
  tests and fix RUN line and checks.

frederik-h committed:
fixup! Add MachineRegisterInfo::getOneNonDBGUse{,r}

frederik-h committed:
Simplify SDWASrcOperand::potentialToConvert

frederik-h committed:
Merge remote-tracking branch 'upstream/main' into sdwa-subreg

frederik-h committed:
Add new test for multiple reg uses and move all tests

frederik-h committed:
Add some comments

frederik-h committed:
Remove unused includes

frederik-h committed:
fixup! Add new test for multiple reg uses and move all tests

arsenm commented on llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp:
```diff
@@ -418,10 +513,36 @@ MachineInstr *SDWASrcOperand::potentialToConvert(const SIInstrInfo *TII,
+  if (isSameReg(LHS, RHS))
+    return true;
+
+  const MachineOperand *Def = findSingleRegDef(&LHS, MRI);
+  const MachineInstr *MI = Def ? Def->getParent() : nullptr;
+
+  // TODO Handle other copy-like instructions?
+  if (!MI || !MI->isRegSequence())
+    return false;
+
+  auto SubReg = regSequenceFindSubreg(*MI, RHS.getReg());
+  return SubReg && LHS.getSubReg() == SubReg;
```
If all you're doing is looking through copy-like instructions, perhaps this pass should not be directly responsible for this. I think either peephole-opt or SIFoldOperands should be pulling the subregister read through the reg_sequence. This change isn't core to the SDWA problem.


i.e.

```
    %4:vreg_64 = REG_SEQUENCE %1, %subreg.sub0, %3, %subreg.sub1
    %7:vreg_64 = REG_SEQUENCE %6, %subreg.sub0, %3, %subreg.sub1
    %8:vgpr_32, %9:sreg_64_xexec = V_ADD_CO_U32_e64 %4.sub0, %7.sub0, 0, implicit $exec
```

Should have been rewritten to directly read the reg_sequence inputs in the first place 

```
    %8:vgpr_32, %9:sreg_64_xexec = V_ADD_CO_U32_e64 %1, %6, 0, implicit $exec
```
